[FLINK-5414] [table] Bump up Calcite version to 1.11

wuchong opened the pull request:
Thanks for contributing to Apache Flink. Before you open your pull request, please take the following check list into consideration.
If your changes take all of the items into account, feel free to open your pull request. For more information and/or questions please refer to the [How To Contribute guide](http://flink.apache.org/how-to-contribute.html).
In addition to going through the list, please provide a meaningful description of your changes.

- [x] General
  - The pull request references the related JIRA issue ("[FLINK-XXX] Jira title text")
  - The pull request addresses only one issue
  - Each commit in the PR has a meaningful commit message (including the JIRA id)

- [ ] Documentation
  - Documentation has been added for new functionality
  - Old documentation affected by the pull request has been updated
  - JavaDoc for public methods has been added

- [x] Tests & Build
  - Functionality added by the pull request is covered by tests
  - `mvn clean verify` has been executed successfully locally or a Travis build has passed


This PR upgrade Calcite to version 1.11. But there are a lot of compatibility issues. I fixed them. Correct me if the way of fixing is wrong.

fhueske commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/table/codegen/ExpressionReducer.scala:
```diff
@@ -106,8 +106,16 @@ class ExpressionReducer(config: TableConfig)
           reducedValues.add(unreduced)
         case _ =>
+          val reducedValue = reduced.getField(reducedIdx)
+          // RexBuilder handle boolean literal incorrectly, convert it into BigDecimal manually
```
Can you check the comment? Shouldn't "boolean" be "double"?

wuchong replied:
Yes, it should be `double`

fhueske commented:
@twalthr, you might want to have a look at the `FlinkTypeFactory` changes

haohui commented:
Looks good to me overall. One question -- I wonder, does it mean that all array types become nullable after this change?

wuchong commented:
Hi @fhueske , yes, Calcite forces an Calc after each aggregate that only renames fields, because we rename every aggregates in Table API which is not necessary. I changed the logic of getting projections on aggregates to only rename the duplicate aggregates. And that works good, no more Calc appended.

Hi @haohui , the ArrayRelDataType is still NOT NULL. I reverted [that line](https://github.com/apache/flink/blob/master/flink-libraries/flink-table/src/main/scala/org/apache/flink/table/calcite/FlinkTypeFactory.scala#L121) which is not need to be changed in this PR.

Cheers,
Jark Wu

fhueske commented:
Thanks for the update @wuchong.
PR looks good to me. Feel free to merge :-)

twalthr commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/table/plan/ProjectionTranslator.scala:
```diff
@@ -108,62 +109,73 @@ object ProjectionTranslator {
-  private def replaceAggregationsAndProperties(
+    val projectNames: mutable.HashSet[String] = new mutable.HashSet[String]
+
+    def replaceAggregationsAndProperties(
```
Can you rename this or the outer function? Having two functions with the same names is confusing.

twalthr commented on flink-libraries/flink-table/src/test/scala/org/apache/flink/table/ExpressionReductionTest.scala:
```diff
@@ -155,15 +155,15 @@ class ExpressionReductionTest extends TableTestBase {
       batchTableNode(0),
       term("select",
-        "13 AS _c0",
+        "CAST(13) AS _c0",
```
Do you know why there are so many unnecessary casts? Is it because of the different nullability?

wuchong replied:
Yes, because I changed the default nullable to `true`, but the reduced constant is `NOT NULL`, so a `CAST` is here. Do you have any ideas to fix this? 

The default nullable changed to `true` is because `UserDefinedScalarFunctionTest.testResults` and `ArrayTypeTest.testArrayLiterals` fail.

twalthr replied:
I will have a look at it again. In general, the only real solution is finally fix FLINK-5177.

haohui replied:
Is it possible to not changing the default nullability while adopting Calcite 1.11? Let me try it out as well.

haohui replied:
Just played around a little bit. I think the problem is that the advanced types are not properly canonized. Using the following diff can pass all tests in `ArrayTypeTest`:

```
--- a/flink-libraries/flink-table/src/main/scala/org/apache/flink/table/calcite/FlinkTypeFactory.scala
+++ b/flink-libraries/flink-table/src/main/scala/org/apache/flink/table/calcite/FlinkTypeFactory.scala
@@ -133,12 +133,18 @@ class FlinkTypeFactory(typeSystem: RelDataTypeSystem) extends JavaTypeFactoryImp
   override def createTypeWithNullability(
     relDataType: RelDataType,
     nullable: Boolean)
-  : RelDataType = relDataType match {
-    case composite: CompositeRelDataType =>
-      // at the moment we do not care about nullability
-      composite
-    case _ =>
-      super.createTypeWithNullability(relDataType, nullable)
+  : RelDataType = {
+    val t = relDataType match {
+      case composite: CompositeRelDataType =>
+        // at the moment we do not care about nullability
+        composite
+      case array: ArrayRelDataType =>
+        val elementType = canonize(createTypeWithNullability(array.getComponentType, nullable))
+        new ArrayRelDataType(array.typeInfo, elementType, nullable)
+      case _ =>
+        super.createTypeWithNullability(relDataType, nullable)
+    }
+    canonize(t)
   }
 }
```

GroupWindowTest is still failing as it misses an identity projection. I'm wondering why `ProjectRemoveRule.INSTANCE` did not kick in...

docete commented:
@wuchong @haohui Could u help to check FLINK-6173 ?
